fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -0,0 +1,198 @@
+
+
+def _parameter_rrefs(module: rpc.RRef) -> List[rpc.RRef]:
+    return [rpc.RRef(p) for p in module.to_here().parameters()]
+
+
+def rloss(loss_func: Callable, input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+    if BOUNCE_TENSORS:
+        return loss_func(input_rref.remote().cpu().to_here(), target_rref.remote().cpu().to_here())
+    else:
+        return loss_func(input_rref.to_here(), target_rref.to_here())
+
+
+def DistributedLoss(loss: nn.Module, *args: Tuple, **kwargs: Dict) -> Callable:
+    loss_func = loss(*args, **kwargs)
+
+    def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+        return rpc.remote(input_rref.owner(), rloss, args=(loss_func, input_rref, target_rref))
+
+    return dloss
+
+
+class MultiProcessPipe(Module):
+    """Wraps an arbitrary :class:`nn.Sequential <torch.nn.Sequential>` module
+    to train on Pipe_. If the module requires lots of memory, Pipe will be
+    very efficient.
+    ::
+
+        model = nn.Sequential(a, b, c, d)
+        model = Pipe(model, balance=[1, 1, 1, 1], chunks=8)
+        output = model(input)
+
+    .. _Pipe: https://arxiv.org/abs/1811.06965
+
+    Pipe combines pipeline parallelism with checkpointing to reduce peak
+    memory required to train while minimizing device under-utilization.
+
+    You should determine the balance when defining a :class:`Pipe` module, as
+    balancing will not be done automatically. The module will be partitioned
+    into multiple devices according to the given balance. You may rely on
+    heuristics to find your own optimal configuration.
+
+    Args:
+        module (torch.nn.Sequential):
+            sequential module to be parallelized
+        balance (ints):
+            list of number of layers in each partition
+
+    Keyword Args:
+        devices (iterable of devices):
+            devices to use (default: all CUDA devices)
+        chunks (int):
+            number of micro-batches (default: ``1``)
+        checkpoint (str):
+            when to enable checkpointing, one of ``'always'``,
+            ``'except_last'``, or ``'never'`` (default: ``'except_last'``)
+        deferred_batch_norm (bool):
+            whether to use deferred BatchNorm moving statistics (default:
+            :data:`False`, see :class:`Deferred Batch Normalization <DeferredBatchNorm>` for more
+            details)
+
+    Raises:
+        TypeError:
+            the module is not a :class:`nn.Sequential <torch.nn.Sequential>`.
+        ValueError:
+            invalid arguments, or wrong balance
+        IndexError:
+            the number of devices is fewer than the number of partitions.
+
+    """
+
+    #: The number of micro-batches.
+    chunks: int = 1
+
+    #: The checkpoint mode to determine when to enable checkpointing. It is one
+    #: of ``'always'``, ``'except_last'``, or ``'never'``.
+    checkpoint: str = "never"
+
+    def __init__(
+        self,
+        module: List[LayerSpec],
+        *,
```

Comment:
Is this right? what does a single * mean?

Reply:
Everything after the '*' must be passed as a keyword argument.